apps/system/js/audio_channel_manager.js

```diff
@@ -0,0 +1,657 @@
+      vibration.onsuccess = () => {
+        this._isVibrateEnabled = vibration.result['vibration.enabled'];
+      };
+      vibration.onerror = () => {
+        this.debug('Error of getting vibration.enabled: ' + vibration.error);
+      };
+      settings.addObserver('vibration.enabled', (evt) => {
+        this._isVibrateEnabled = evt.settingValue;
+      });
+      this.debug('Start the audio channel manager module.');
+    },
+
+    /**
+     * Add an app into manager to manage its audio channels.
+     *
+     * @param {AppWindow} app An app you want to manage.
+     */
+    registerAudioChannel: function(app) {
+      var allowedAudioChannels = app.browser.element.allowedAudioChannels;
+      // Feature detection for using the new audio channel management module.
+      allowedAudioChannels && allowedAudioChannels.forEach((channel) => {
+        var appId = app.instanceID;
+        var channelName = channel.name;
+        channel.onactivestatechanged = () => {
+          this.debug('Active state is changed',
+              'app id: '+ appId, 'app name: ' + app.name,
+              'channel name: ' + channelName);
+          channel.isActive().onsuccess = (evt) => {
+            var isActive = evt.target.result;
+            this.debug('The active state is ' + isActive);
+            if (isActive) {
+              this._handleAudioChannel('competition', appId, channel);
+            } else {
+              this._resetAudioChannel(appId, channelName);
+              this._resetFadedAudioChannelBy(appId, channelName);
+              this._resumeInterruptedAudioChannelBy({
+                interruptionAppAduioChannel: {
+                  appId: appId,
+                  channel: channel
+                }
+              });
+            }
+          };
+        };
+        this.debug(app.name + ' registers ' + channelName + ' channel');
+      });
+    },
+
+    /**
+     * Remove an app from manager,
+     * if you don't wnat to manage it anymore.
+     *
+     * @param {AppWindow} app An app you want to remove.
+     */
+    unregisterAudioChannel: function(app) {
+      var allowedAudioChannels = app.browser.element.allowedAudioChannels;
+      // Feature detection for using the new audio channel management module.
+      allowedAudioChannels && allowedAudioChannels.forEach((channel) => {
+        channel.onactivestatechanged = undefined;
+        this.debug(app.name + ' unregisters ' + channel.name + ' channel');
+      });
+    },
+
+    /**
+    * Handle audio channels when the app is opened or closed.
+    *
+    * @param {DOMEvent} evt The event.
+    */
+    handleEvent: function(evt) {
+      var appId = evt.detail.instanceID;
+      var channel;
+      var index;
+      switch (evt.type) {
+        case 'appopened':
+          this.debug(appId + ' is opened');
+          this._resumeInterruptedAudioChannelBy({ opendAppId: appId });
+          break;
+
+        case 'appclosed':
```

Comment:
BTW....I don't think you should use appopened/appclosed...
Again, your case does not cover attention window.

Please listen to hierarchytopmostwindowchanged event insteadly and use Service.query('getTopMostWindow') to know who is top most right now.